### PR TITLE
Revise how modules imports are done to better support `import public`

### PR DESCRIPTION
**Pull request by thomasvl**
Historically Swift didn't have a good way to model `import public` in .proto
files. But by using `@_exported import`, we can get the majority of the way
there. This makes restructuring of proto files less likely to break Swift source
code; meaning Swift is less of a "special case" compared to other languages.

If generating with `public` or `package` visibility, then instead of importing
the whole module of for an `import public`, explicitly re-export the imports of
the types from that file. That will make them usable for the given source file
but will also let the types continue to be used by someone just importing this
module.

This also means a file with no `import public` usages becomes must simpler as it
can just rely on its direct dependencies to provide all the types that could be
used.

### CODE REVIEW

**Comment by thomasvl** [comment]
I split the setup of the compile test infra into the first comment, so the actually changes for `import public` are better isolated to hopefully make reviewing a comment at a time better to see the specific changes.

**Comment by thomasvl** [comment]
fyi @FranzBusch @Lukasa @allevato

**Comment by tbkka** [comment]
The new `CompileTest` looks intriguing.

**Review by allevato — APPROVED** [review]

File: Sources/protoc-gen-swift/Descriptor+Extensions.swift [hunk 1]
```diff
@@ -14,6 +14,150 @@ extension FileDescriptor {
...
+  // like anyone else would need the logic. Swift GRPC support probably stick
+  // with the support for the module mappings.
+  public func computeImports(
+    namer: SwiftProtobufNamer,
+    reexportPublicImports: Bool,
+    asImplementationOnly: Bool
+  ) -> String {
+    // The namer should be configured with the module this file generated for.
+    assert(namer.targetModule == (namer.mappings.moduleName(forFile: self) ?? ""))
+    // Booth options can't be enabled.
```
[the remark this answers is not in the document]
Comment (thomasvl) [reply Sources/protoc-gen-swift/Descriptor+Extensions.swift]: Fixed.

**Comment by thomasvl** [comment]
> The new `CompileTest` looks intriguing.

This could all just be in the main `Package.swift`, but I didn't like the idea that it impacted everyone just trying to use swift-protobuf, that's mainly why I spun it out on its own.  We can always revisit that in the future.  We could also consider combining the directory with the _PluginExample_ one; but since that's both a _test_ and an _example_, I didn't do that to start with.

**Review by FranzBusch — COMMENTED** [review]
I like the changes here. The only thing I would like us to reconsider is checking in yet more generated proto files. In the `PluginExamples`, directory I have already used the Swift PM plugin to generate Swift files on the fly and it also does a compile check. The CI for that is already setup correctly and works.

Maybe we can merge the two setups into one since they are aiming to do the same thing. The added benefit of that would be that we make sure everything is doable via the Swift PM plugin as well.

**Comment by thomasvl** [comment]
> I like the changes here. The only thing I would like us to reconsider is checking in yet more generated proto files. In the `PluginExamples`, directory I have already used the Swift PM plugin to generate Swift files on the fly and it also does a compile check. The CI for that is already setup correctly and works.

I tried to cover this in CompileTests/MultiModule/README.md -

> ...
> This can't use the SwiftPM Plugin as that currently doesn't have support for the module mappings.

We have the request for this in #1450, but I thought the opinion was to not do this and instead wait on the hope that SwiftPM would eventually expose something so the SwiftPM plugin could eventually generate this info itself.

> Maybe we can merge the two setups into one since they are aiming to do the same thing. The added benefit of that would be that we make sure everything is doable via the Swift PM plugin as well.

Since we don't support the mappings, I don't believe a developer could create such a set up.

**Comment by FranzBusch** [comment]
> I tried to cover this in CompileTests/MultiModule/README.md -

Oh sorry I missed that part. Makes sense then! Thanks for clarifying.

**Review by FranzBusch — APPROVED** [review]

**Comment by thomasvl** [comment]
> The new `CompileTest` looks intriguing.

@tbkka did you want to look at this PR in a little more detail, or am I good to merge it?

**Review by tbkka — APPROVED** [review]

**Comment by tbkka** [comment]
If Franz and Tony are both okay with it, then please go ahead!

**Pull request merged**